peepingtom/datablocks/simpleblocks/imageblock.py

```diff
@@ -1,38 +1,61 @@
+import logging
+
 import numpy as np
 from xarray import DataArray
 
 from .simpleblock import SimpleBlock
 from ...depictors import ImageDepictor
 
 
+logger = logging.getLogger(__name__)
+
+
 class ImageBlock(SimpleBlock):
     """
     n-dimensional image block
     data can be interpreted as n-dimensional images
     """
     _depiction_modes = {'default': ImageDepictor}
 
-    def __init__(self, data=(), pixel_size=None, **kwargs):
-        super().__init__(data, **kwargs)
+    def __init__(self, data=(), pixel_size=None, shape=None, ndim=None, **kwargs):
         # TODO this is a workaround until napari #2347 is fixed
-        if pixel_size is None:
-            pixel_size = np.ones(self.ndim)
-        self.pixel_size = pixel_size
+        self.pixel_size = pixel_size  # no checking here, or we screw up lazy loading
+        self._shape = shape
+        self._ndim = ndim
+        super().__init__(data, **kwargs)
 
     def _data_setter(self, data):
         data = np.asarray(data)  # asarray does not copy unless needed
         if data.ndim < 2:
             raise ValueError('images must have at least 2 dimensions')
         dims = ('z', 'y', 'x')
-        return DataArray(data, dims=dims[-data.ndim:])
+        data = DataArray(data, dims=dims[-data.ndim:])
+
+        # set pixel_size if needed
+        if self.pixel_size is None or np.all(self.pixel_size == 0):
+            pixel_size = np.ones(data.ndim)
+            pixel_size = np.broadcast_to(pixel_size, data.ndim)
+            self.pixel_size = pixel_size
+
+        return data
 
     @property
     def ndim(self):
+        if callable(self._data):
+            return self._ndim
         return self.data.ndim
 
     @property
     def dims(self):
+        if callable(self._data):
+            return len(self._shape)
         return self.data.dims
 
+    @property
+    def shape(self):
+        if callable(self._data):
```

Comment:
seems counterintuitive that a null value would be returned when a lazy imageblock was being worked with... how does the 'user' know if the data is a placeholder/null value being returned?

Reply:
I'm not sure I get what you mean by "null" value. These properties (shape, dims...) are supposed to be set by the reader when making a lazy datablock. This is because ImageBlocks *need* some information to be initialized correctly (we want to check that the shape makes sense, and sometimes we need it to rescale particles). This means that currently one cannot make a lazy datablock and not provide its shape, or things will fail.

I don't really like it, but so far I haven't come up with a nicer solution than enforcing what the readers should provide. I think that creating datablocks without even knowing if the data is image-like and then risking failing once the data gets loaded is a worse solution than this... Thoughts?

Reply:
Okay I'm with you - definitely agree that breaking later is worse

I didn't see that it was enforced to set these properties correctly for a lazy block, I just saw that these would be defaulted if not provided and worried about providing those default values - is it enforced or it's just a "I know I should do this" situation when implementing a lazy block?

Reply:
They are not really enforced, you're right. Maybe they should be? In that case, it probably would be better to have a differenty interface for lazy constructors, like you were suggesting... I'm not a huge fan of either of these options, I have to say :P I'll think about it!